### PR TITLE
[3.10] Add a check whether we are running the core backend template and issue an info when that is not the case

### PR DESCRIPTION
**Pull request by zero-24**
### Summary of Changes

Add a check whether we are running the core tempaltes where the backend it required and the frontend is recommended

### Testing Instructions

- Install 3.10.0
- install a non-core backend template (https://www.joomlatema.net/joomla-templates/free-joomla-templates/adminim-free-joomla-administrator-template.html)
- set the template as default for the backend
- apply this patch
- point the update server to joomla 4 via the Joomla Next update server
- notice the new message

### Actual result BEFORE applying this Pull Request

No checks for the default templates

image

### Expected result AFTER applying this Pull Request

image

### Documentation Changes Required

https://docs.joomla.org/Pre-Update_Check - Has to be updated

TTs have to be informed about the language stirngs

cc @joomla/cms-release-team @HLeithner @bembelimen

### CODE REVIEW

**Comment by brianteeman** [comment]
These are NOT php or database settings

Core backend template is NOT a requirement

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 33a0aed794ea8d2cc98b75532dc45e3623292ddf<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35107">issues.joomla.org/tracker/joomla-cms/35107</a>.</sub>

**Comment by zero-24** [comment]
> Core backend template is NOT a requirement

Will move it to the optional thing than fine for me.

> These are NOT php or database settings

Agree its on my list to be updated too.

**Comment by zero-24** [comment]
The requested changes have been pushed. They are now both under "the seccond box" and the title has been updated too.

**Comment by brianteeman** [comment]
why check for isis? there is more than 1 admin template

File: administrator/components/com_joomlaupdate/models/default.php [hunk 1]
```diff
@@ -1223,6 +1223,40 @@ public function getPhpOptions()
 		return $options;
 	}
 
+	/**
+	 * Checks whether the default backend tempalte is isis
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * Checks if the default backend template is isis
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 2]
```diff
@@ -1223,6 +1223,40 @@ public function getPhpOptions()
...
+		if ($this->isTemplateActive('isis'))
+		{
+			return true;
+		}
+
+		return false;
+	}
+
+	/**
+	 * Checks whether the default frontend tempalte is protostar
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * Checks if the default frontend template is protostar
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 3]
```diff
@@ -1791,4 +1850,59 @@ protected function translateExtensionName(&$item)
 		// Translate the extension name if possible
 		$item->name = strip_tags(JText::_($item->name));
 	}
+
+	/**
+	 * Checks whether an given template is active
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * Checks if a given template is active
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 4]
```diff
@@ -1791,4 +1850,59 @@ protected function translateExtensionName(&$item)
...
+	 *
+	 * @param   string  $template  The tempalte name to be checked
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * @param   string  $template  The template name to be checked
```

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 5]
```diff
@@ -116,9 +120,9 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_PACKAGE="Update package URL"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PACKAGE_REINSTALL="Reinstall package URL"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_PREUPDATE_CHECK="Pre-Update Check for Joomla %s"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED="Recommended"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED_SETTINGS_PASSED="Recommended PHP Settings : Passed"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED_SETTINGS_WARNING="Recommended PHP Settings : Warning"
-COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED_SETTINGS_DESC="These settings are recommended for PHP in order to ensure full compatibility with Joomla. However, Joomla! will still operate if your settings do not quite match the recommended configuration."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED_SETTINGS_PASSED="Recommended PHP & Joomla Settings : Passed"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED_SETTINGS_WARNING="Recommended PHP & Joomla Settings : Warning"
+COM_JOOMLAUPDATE_VIEW_DEFAULT_RECOMMENDED_SETTINGS_DESC="These settings are recommended for PHP and Joomla in order to ensure full compatibility. However, Joomla! will still upgrade if your settings do not quite match the recommended configuration."
```
Comment (brianteeman) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: please alpha sort

**Comment by brianteeman** [comment]
Neither of these checks are anything to do with compatibility. They are only really needed to perhaps make the update easier. This is just creating **_FUD_**

**Comment by richard67** [comment]
Here in the 4.0 update SQL we check for hathor and isis and update to atum if one of these is used: https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-03-05.sql#L23

And one row below we check for protostar and beez 3 for the frontend.

So why do the language strings added by this PR here not mention hathor or beez3? It should work same well as with the other core templates.

We should not provide false information in language strings which are made to guide people.

**Comment by brianteeman** [comment]
This and the other PR should be closed and re-evaluated. There is ZERO reason for either PR. 

Not withstanding the code issues the concept of both PR is wrong!! They only succeed in spreading fear, uncertainty and doubt about the upgrade process. 

I can't believe you are even remotely thinking about such a change **after** the final release candidate.

**Comment by brianteeman** [comment]
I have tested this item :red_circle: unsuccessfully on 164496e012fa6133177a460f85cdc62812735fd7<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35107">issues.joomla.org/tracker/joomla-cms/35107</a>.</sub>

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini (hunk 5 above)
Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: https://github.com/joomla/joomla-cms/pull/35107/commits/4509aaca09173306adad4039a1afd8e0df2ef078

**Comment by zero-24** [comment]
> So why do the language strings added by this PR here not mention hathor or beez3? It should work same well as with the other core templates.

Isnt hathor gone already? I can add them for sure.

> I can't believe you are even remotely thinking about such a change after the final release candidate.

It has been raised by the @joomla/cms-release-team that there should be messages like that. I see where an "pre upgrade checker" could warn about such issues that affect the site performance after the upgrade or the upgrade process.

For example an non-core backed templates or non-core frotend template could break the site once upgraded so the recommendation would be for the time of the upgrade to switch to the core where we know that the upgrade works and they are beeing replaced by 4.x compatible code than you can go and enable the other stuff step by step too.

The router stuff is something different happening under the hood so the idea was to tell the people that there could be changes in the URLs joomla generates after the upgrade.

**Comment by brianteeman** [comment]
> Isnt hathor gone already? 

No  - just not supported. You obviously didnt even check. https://github.com/joomla/joomla-cms/tree/3.10-dev/administrator/templates

> It has been raised by the @joomla/cms-release-team that there should be messages like that. 

Excuse my language but "a bit bloody late"

> For example an non-core backed templates or non-core frotend template could break the site once upgraded so the recommendation would be for the time of the upgrade to switch to the core where we know that the upgrade works and they are beeing replaced by 4.x compatible code than you can go and enable the other stuff step by step too.

But that is not what you are doing or saying with these pr

> The router stuff is something different happening under the hood so the idea was to tell the people that there could be changes in the URLs joomla generates after the upgrade.

Again that is not what is being done or said with these PR.

**Comment by zero-24** [comment]
> No - just not supported. You obviously didnt even check. https://github.com/joomla/joomla-cms/tree/3.10-dev/administrator/templates

Hmm I'm sure we did a postinstall for hathor users but yes seems its still shipped.

> But that is not what you are doing or saying with these pr

Than its great that you raised this questions so that this can be clarified. What can be done to do what I'm trying to do?

File: administrator/components/com_joomlaupdate/models/default.php [hunk 6]
```diff
@@ -1791,4 +1850,59 @@ protected function translateExtensionName(&$item)
...
+			)
+		)->from(
+			$db->qn('#__template_styles')
+		)->where(
+			$db->qn('template') . ' = ' . $db->q($template)
+		);
+
+		$templates = $db->setQuery($query)->loadObjectList();
+
+		$home = array_filter($templates, function($value) {
```
Comment (zero-24) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
		$home = array_filter($templates, function($value)
		{
```

**Comment by richard67** [comment]
Beside all other points: Now as both checks are located in the “Recommended …” ckecks, the title of this PR is wrong regarding required and recommended.

**Comment by zero-24** [comment]
Fixed thanks

**Comment by HLeithner** [comment]
Maybe brian is partly right. The way to fix this is to complex, reduce it to a simple change in https://github.com/joomla/joomla-cms/blob/628aee40e1ebe9012458e8d27a2191ae399e7938/administrator/components/com_admin/sql/updates/mysql/4.0.0-2018-03-05.sql#L22-L24

Set `atum` to the default template on upgrade and leave the frontend template alone, the frontend template can't break the admin on update but the backend template can. Activating the old template again by the user is his/her decision. And I would expect less support requests at least for copy/paste admin templates.

**Comment by zero-24** [comment]
> Set atum to the default template on upgrade and leave the frontend template alone, the frontend template can't break the admin on update but the backend template can. 

You mean ignoring any possible non-core backend template after the upgrade and force everyone to atum right?

**Comment by HLeithner** [comment]
> > Set atum to the default template on upgrade and leave the frontend template alone, the frontend template can't break the admin on update but the backend template can.
> 
> You mean ignoring any possible non-core backend template after the upgrade and force everyone to atum right?

yes

**Comment by zero-24** [comment]
> > > Set atum to the default template on upgrade and leave the frontend template alone, the frontend template can't break the admin on update but the backend template can.
> > 
> > 
> > You mean ignoring any possible non-core backend template after the upgrade and force everyone to atum right?
> 
> yes

Hmm with or without the warning mentiond here? Maybe change the message something like "hey you are one a non-core backend template that will be changed to atum after the upgrade"?

**Comment by HLeithner** [comment]
Don't think it's needed, my opinion is that people would expect a new backend template.

**Comment by zero-24** [comment]
> Don't think it's needed, my opinion is that people would expect a new backend template.

Ok so the way to go would be to drop this PR but change the update SQL in 4.0 to always update the backend template. Fine for me.

**Comment by HLeithner** [comment]
@wilsonge if this is ok for you then yes PR should be done against 4.0

**Comment by brianteeman** [comment]
@HLeithner is correct with the way to go. I logged in this morning to say virtually the same thing. The only difference being that I thought (maybe wrong) that we did already set atum as the default admin template.

I also wanted to point out that just because someone is using a non core admin template does not automatically mean that its not compatible with joomla 4. Making that assumption is the same as telling people to disable all extensions without even checking for their compatibility.

The only thing that I might consider doing if people really insist would be a notice that said "We detected that you are not using a core admin template. You might find the upgrade process smoother if you switch to use the Isis template."

**Comment by zero-24** [comment]
> The only thing that I might consider doing if people really insist would be a notice that said "We detected that you are not using a core admin template. You might find the upgrade process smoother if you switch to use the Isis template."

PR updated with that.

**Comment by brianteeman** [comment]
Sorry I wasn't clear. I absolutely would not put the notice in the checks. I would revert all the changes that you have made from php recommendations to php and joomla recommendations and simply do this notice as an on screen message. As I am clearly having problems describing things here is a picture.

image

**Comment by richard67** [comment]
> Sorry I wasn't clear. I absolutely would not put the notice in the checks. I would revert all the changes that you have made from php recommendations to php and joomla recommendations and simply do this notice as an on screen message. As I am clearly having problems describing things here is a picture.

@zero-24 I agree with @brianteeman .

**Comment by zero-24** [comment]
code has been updated @brianteeman @richard67

**Comment by zero-24** [comment]
Ah missed that there is also a link requested to the styles will update the language string now

File: administrator/components/com_joomlaupdate/models/default.php [hunk 7]
```diff
@@ -1791,4 +1791,63 @@ protected function translateExtensionName(&$item)
 		// Translate the extension name if possible
 		$item->name = strip_tags(JText::_($item->name));
 	}
+
+	/**
+	 * Checks whether an given template is active
+	 *
+	 * @param   string  $template  The tempalte name to be checked
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * @param   string  $template  The template name to be checked
```

File: administrator/components/com_joomlaupdate/models/default.php [hunk 8]
```diff
@@ -1791,4 +1791,63 @@ protected function translateExtensionName(&$item)
 		// Translate the extension name if possible
 		$item->name = strip_tags(JText::_($item->name));
 	}
+
+	/**
+	 * Checks whether an given template is active
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * Checks whether a given template is active
```

**Comment by zero-24** [comment]
language string & screenshots updated now too.

File: administrator/components/com_joomlaupdate/views/default/tmpl/default_preupdatecheck.php [hunk 9]
```diff
@@ -44,6 +44,16 @@
 	)
 );
 
+if (version_compare($this->updateInfo['latest'], '4', '>=') && $this->isBackendTeamplateIsis === false)
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/views/default/tmpl/default_preupdatecheck.php]: typo template not teamplate

File: administrator/components/com_joomlaupdate/views/default/view.html.php [hunk 10]
```diff
@@ -102,6 +102,7 @@ public function display($tpl = null)
 		$this->phpSettings            = $model->getPhpSettings();
 		$this->nonCoreExtensions      = $model->getNonCoreExtensions();
 		$this->nonCoreCriticalPlugins = $model->getNonCorePlugins(array('system','user','authentication','actionlog','twofactorauth'));
+		$this->isBackendTeamplateIsis = (bool) $this->getModel()->isTemplateActive('isis');
```
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/views/default/view.html.php]:
```suggestion
		$this->isBackendTemplateIsis = (bool) $this->getModel()->isTemplateActive('isis');
```

File: administrator/components/com_joomlaupdate/views/default/tmpl/default_preupdatecheck.php (hunk 9 above)
Comment (brianteeman) [inline administrator/components/com_joomlaupdate/views/default/tmpl/default_preupdatecheck.php]:
```suggestion
if (version_compare($this->updateInfo['latest'], '4', '>=') && $this->isBackendTemplateIsis === false)
```

File: administrator/components/com_joomlaupdate/views/default/view.html.php [hunk 11]
```diff
@@ -102,7 +102,7 @@ public function display($tpl = null)
 		$this->phpSettings            = $model->getPhpSettings();
 		$this->nonCoreExtensions      = $model->getNonCoreExtensions();
 		$this->nonCoreCriticalPlugins = $model->getNonCorePlugins(array('system','user','authentication','actionlog','twofactorauth'));
-		$this->isBackendTeamplateIsis = (bool) $this->getModel()->isTemplateActive('isis');
+		$this->isBackendTemplateIsis = (bool) $this->getModel()->isTemplateActive('isis');
```
Comment (richard67) [inline administrator/components/com_joomlaupdate/views/default/view.html.php]:
```suggestion
		$this->isBackendTemplateIsis  = (bool) $model->isTemplateActive('isis');
```
Align the "=" with those above

Comment (richard67) [reply administrator/components/com_joomlaupdate/views/default/view.html.php]: Ah, and why don't you just use `$model` instead of `$this->getModel()`?

**Comment by richard67** [comment]
@zero-24 This PR has a conflict now due to previously merged PR #35105 . Please resolve. Thanks in advance.

**Comment by zero-24** [comment]
merge confilct fixed.

**Comment by HLeithner** [comment]
Sorry but the warning message position is completely the wrong place for it in my opinion.

File: administrator/components/com_joomlaupdate/models/default.php [hunk 12]
```diff
@@ -1800,4 +1800,63 @@ protected function translateExtensionName(&$item)
 		// Translate the extension name if possible
 		$item->name = strip_tags(JText::_($item->name));
 	}
+
+	/**
+	 * Checks whether a given template is active
+	 *
+	 * @param   string  $template  The template name to be checked
+	 *
+	 * @return  bool
```
Comment (PhilETaylor) [inline administrator/components/com_joomlaupdate/models/default.php]:
```suggestion
	 * @return  boolean
```

**Comment by zero-24** [comment]
I have finally found a custom backend template to test: https://www.joomlatema.net/joomla-templates/free-joomla-templates/adminim-free-joomla-administrator-template.html

**Comment by zero-24** [comment]
I have just updated the test instructions. Would be great to get some tests here with the new blue message.

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini [hunk 13]
```diff
@@ -79,6 +79,7 @@ COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_WARNING_UNKNOWN="Unkown error"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSION_SERVER_ERROR="Update Server error"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS="Extensions Pre-Update Check"
 COM_JOOMLAUPDATE_VIEW_DEFAULT_EXTENSIONS_NONE="No extensions installed."
+COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_BACKEND_TEMPLATE_USED_NOTICE="We detected that you are not using a core admin template. You might find the upgrade process smoother if you <a href="_QQ_"index.php?option=com_templates&view=styles&client_id=1"_QQ_">switch</a> to use the Isis template."
```
Comment (conconnl) [inline administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: To my knowledge `_QQ_` should not be used anymore.
I was told that `\"` is the correct way nowadays.

Comment (brianteeman) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: this is a j3 string not j4

Comment (conconnl) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: J3 also supports `\"` and already uses that on a lot of places, which makes it compatible with J4 as well.

Comment (zero-24) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]:
```suggestion
COM_JOOMLAUPDATE_VIEW_DEFAULT_NON_CORE_BACKEND_TEMPLATE_USED_NOTICE="We detected that you are not using a core admin template. You might find the upgrade process smoother if you <a href='index.php?option=com_templates&view=styles&client_id=1'>switch</a> to use the Isis template."
```

That should do the trick too right?

**Review by conconnl — APPROVED** [review]

**Review by conconnl — APPROVED** [review]

File: administrator/language/en-GB/en-GB.com_joomlaupdate.ini (hunk 13 above)
Comment (conconnl) [reply administrator/language/en-GB/en-GB.com_joomlaupdate.ini]: I would think so comparing it to other strings I have seen.

**Review by conconnl — APPROVED** [review]

**Comment by ChristineWk** [comment]
Since I saw this PR under Milestones 3.10.2, I wanted to try it & to check the blue message :-) 
Joomla Version: 3.10.1

Downloaded to install admin template (Adminim_Joomla_Admin_Template.zip) but got:
Error
There was an error uploading this file to the server.
Unable to find install package<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35107">issues.joomla.org/tracker/joomla-cms/35107</a>.</sub>

**Comment by ChristineWk** [comment]
Update to above: Will try another one (unzip first) .....

**Comment by ChristineWk** [comment]
- 2nd admin template: Installed unzipped via FTP. Was not visible in the backend.
- 3rd admin template: Vallis installed & activated. 
screen shot 2021-09-08 at 21 08 29
- applied this patch
- pointed the update server to joomla 4 via the Joomla Next update server

Pre-Update Check for Joomla 4.0.2

Result:
screen shot 2021-09-08 at 21 09 59

Will leave it for today, but want to go back later to Isis and J 3.10.1 again :-)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35107">issues.joomla.org/tracker/joomla-cms/35107</a>.</sub>

**Comment by zero-24** [comment]
> Will leave it for today, but want to go back later to Isis and J 3.10.1 again :-)

The message seems to be correct to me, so this would be a successfull test. :)

**Comment by ChristineWk** [comment]
I have tested this item :white_check_mark: successfully on ad7b419aee0a817bc57521052885492f265e1a4c<br /><br />But it's not in blue :-)
Would also prefer to switch to Isis before updating. Similar to the frontend template on Protostar (if you were on a copy)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/35107">issues.joomla.org/tracker/joomla-cms/35107</a>.</sub>

**Comment by zero-24** [comment]
> But it's not in blue :-)

I have just double checked that, it seems the template does not show "info" as blue. But isis does so it seems the template decided to show information messages in yellow which is a chooise of the template.

**Comment by zero-24** [comment]
> Sorry but the warning message position is completely the wrong place for it in my opinion.

I'm sorry @HLeithner seems I have missed this comment. Do you have another proposal where that message should be shown while it should only be shown too a small number of sites anyway right?

**Comment by HLeithner** [comment]
like all other warnings, maybe in a new section.

**Comment by zero-24** [comment]
> like all other warnings, maybe in a new section.

What kind of new section do you have in mind?

**Comment by zero-24** [comment]
@HLeithner can we get a response here so the PR can be adjusted or closed based on the feedback?

**Comment by HLeithner** [comment]
I thought about something like "Required PHP & Database Settings".

**Comment by zero-24** [comment]
Is it required or an php or database setting? This is exactly the place where all of this started but it was moved outside of this checks for that reason and moved into a standalone blue message.

**Comment by HLeithner** [comment]
not in this box, a new box like this box

**Comment by zero-24** [comment]
And how would you call it, what would we show when you are running isis? Just as a reminder its only showed when you have a non-isis backend template?

**Comment by zero-24** [comment]
Will take this now in as it is for now. When we find better ways to display it I'm open to change it.

**Pull request merged**